fork-off/src/main.rs

```diff
@@ -12,168 +19,336 @@ pub struct Config {
+        let keys: Vec<StorageKey> = self.make_request(body).await;
+        keys
+    }
+
+    async fn get_all_keys(&self, block_hash: BlockHash) -> Vec<StorageKey> {
+        let mut last_key_fetched = None;
+        let mut all_keys = Vec::new();
+        loop {
+            let new_keys = self
+                .get_keys_range(KEYS_BATCH_SIZE, last_key_fetched, Some(block_hash.clone()))
+                .await;
+            all_keys.extend_from_slice(&new_keys);
+            info!(
+                "Got {} new keys and have {} in total",
+                new_keys.len(),
+                all_keys.len()
+            );
+            if new_keys.len() < KEYS_BATCH_SIZE as usize {
+                break;
+            } else {
+                last_key_fetched = Some(new_keys.last().unwrap().clone());
+            }
         }
-        None => serde_json::json!({
-            "jsonrpc": "2.0",
-            "id": 1,
-            "method": "state_getKeysPaged",
-            "params": [ key, 1 ]
-        }),
-    };
+        all_keys
+    }
 
-    let response: Value = reqwest::Client::new()
-        .post(http_rpc_endpoint)
-        .json(&body)
-        .send()
-        .await
-        .expect("Storage request has failed")
-        .json()
-        .await
-        .expect("Could not deserialize response as JSON");
+    async fn get_value(&self, key: StorageKey, block_hash: BlockHash) -> StorageValue {
+        let body = get_storage_value_body(key, Some(block_hash));
+        self.make_request(body).await
+    }
 
-    debug!(target: "fork", "get_key response: {}", response);
+    async fn get_values(
+        &self,
+        keys: Vec<StorageKey>,
+        block_hash: BlockHash,
+        num_workers: u32,
+    ) -> Storage {
+        let n_keys = keys.len();
+        let input = Arc::new(Mutex::new(keys));
+        let output = Arc::new(Mutex::new(HashMap::with_capacity(n_keys)));
+        let mut workers = Vec::new();
 
-    let result = response["result"]
-        .as_array()
-        .expect("No result in response");
+        for id in 0..(num_workers as usize) {
+            workers.push(StateFetcher::value_fetching_worker(
+                self.http_rpc_endpoint.clone(),
+                id,
+                block_hash.clone(),
+                input.clone(),
+                output.clone(),
+            ));
+        }
+        info!("Started {} workers to download values.", workers.len());
+        join_all(workers).await;
+        assert!(input.lock().is_empty(), "Not all keys were fetched");
+        let mut guard = output.lock();
+        std::mem::take(&mut guard)
+    }
 
-    if !result.is_empty() {
-        return Some(
-            result
-                .first()
-                .expect("No key in result")
-                .as_str()
-                .expect("Not a string")
-                .to_owned(),
-        );
+    pub async fn get_full_state(&self, num_workers: u32) -> Storage {
```

Comment:
a) this could be called `get_full_state_at` and take a block hash argument or take an Option<Hash> and default to the  most recent finalized one?

Reply:
Sure, added `get_full_state_at` which takes an option and `get_full_state_at_best_block` that takes no argument.